### PR TITLE
Feature - add Azure Table Storage dependency tracking

### LINKED ISSUE #60: Provide capability to track an Azure Table Storage dependency
**Is your feature request related to a problem? Please describe.**
Provide capability to track an Azure Table Storage dependency.

**Describe the solution you'd like**
We should provide a `LogTableStorageDependency` overload for `ILogger` which is optimized for Azure Table Storage and correctly generates the [Azure Application Insights App Map](https://docs.microsoft.com/en-gb/azure/azure-monitor/app/app-map).

**Describe alternatives you've considered**
None

**Additional context**
Dependency type should be used from [official SDK](https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/WEB/Src/DependencyCollector/Shared/Implementation/RemoteDependencyConstants.cs).

### PR DESCRIPTION
**Pull request by stijnmoreels**
Closes #60

### CODE REVIEW

**Comment by stijnmoreels** [comment]
image

**Comment by stijnmoreels** [comment]
image

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs [hunk 1]
```diff
@@ -33,6 +33,15 @@ public static class ILoggerExtensions
             + ContextProperties.DependencyTracking.IsSuccessful + "} - Context: {@"
             + ContextProperties.EventTracking.EventContext + "})";
 
+        private const string DependencyWithoutDataFormat =
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:36]: Added this bc I didn't come up with a dependency data value. If you think it's usable, we could add also some kind of "operation name" like the SQL dependency to set as dependency data. Then we could reuse the dependency data format.

File: docs/features/writing-different-telemetry-types.md [hunk 2]
```diff
@@ -75,6 +76,44 @@ using (var measurement = DependencyMeasurement.Start())
...
+    { "Collumn", "productID" }
+};
+
+var durationMeasurement = new Stopwatch();
+
+// Start measuring
+durationMeasurement.Start();
+var startTime = DateTimeOffset.UtcNow;
+
+_logger.LogTableStorageDependency(tableName: "orders", isSuccessful: true, startTime, durationMeasurement.Elapsed, telemetryContext);
```
Comment (tomkerkhove) [inline docs/features/writing-different-telemetry-types.md:97]: This should also require the name of the storage account

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:36]: From the looks of it they use the account name as the target so we should pass the table name as data
image
https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-map?tabs=net#composite-application-map

File: docs/features/writing-different-telemetry-types.md (hunk 2 above)
Comment (stijnmoreels) [reply docs/features/writing-different-telemetry-types.md:98]: Ok, added that.

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:36]: Yes, done that.

File: docs/features/writing-different-telemetry-types.md [hunk 3]
```diff
@@ -75,6 +76,45 @@ using (var measurement = DependencyMeasurement.Start())
...
+    { "Order", "ABC" },
+};
+
+var durationMeasurement = new Stopwatch();
+
+// Start measuring
+durationMeasurement.Start();
+var startTime = DateTimeOffset.UtcNow;
+
+_logger.LogTableStorageDependency(tableName: "orders", accountName: "orderAccount", isSuccessful: true, startTime, durationMeasurement.Elapsed, telemetryContext);
```
Comment (tomkerkhove) [inline docs/features/writing-different-telemetry-types.md:98]: I'm going to be a pain, but asking for account and then table makes more sense as it goes from least to most specific

Comment (tomkerkhove) [reply docs/features/writing-different-telemetry-types.md:98]: Also how it's typically done in SDKs/samples, etc.

File: docs/features/writing-different-telemetry-types.md [hunk 4]
```diff
@@ -75,6 +76,45 @@ using (var measurement = DependencyMeasurement.Start())
...
+// Output: "Dependency Azure Table Storage orders named orderAccount in 00:00:00.2521801 at 03/23/2020 09:56:31 +00:00 (Successful: True - Context: [Collumn, productID])"
```
Comment (tomkerkhove) [inline docs/features/writing-different-telemetry-types.md:99]: This needs to be updated

File: docs/features/writing-different-telemetry-types.md [hunk 5]
```diff
@@ -75,6 +76,45 @@ using (var measurement = DependencyMeasurement.Start())
...
+var telemetryContext = new Dictionary<string, object>
+{
+    { "Collumn", "productID" }
+};
+
+// Start measuring
+using (var measurement = DependencyMeasurement.Start())
+{
+    _logger.LogTableStorageDependency(tableName: "orders", accountName: "orderAccount", isSuccessful: true, measurement, telemetryContext);
+    // Output: "Dependency Azure Table Storage orders named orderAccount in 00:00:00.2521801 at 03/23/2020 09:56:31 +00:00 (Successful: True - Context: [Collumn, productID])"
```
Comment (tomkerkhove) [inline docs/features/writing-different-telemetry-types.md:114]: Needs to be updated as well once finished

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs [hunk 6]
```diff
@@ -291,17 +291,53 @@ public static void LogServiceBusDependency(this ILogger logger, string entityNam
...
+        /// </summary>
+        /// <param name="logger">Logger to use</param>
+        /// <param name="tableName">Name of the Table Storage resource</param>
+        /// <param name="accountName">Account of the storage resource</param>
+        /// <param name="isSuccessful">Indication whether or not the operation was successful</param>
+        /// <param name="startTime">Point in time when the interaction with the dependency was started</param>
+        /// <param name="duration">Duration of the operation</param>
+        /// <param name="context">Context that provides more insights on the dependency that was measured</param>
+        public static void LogTableStorageDependency(this ILogger logger, string tableName, string accountName, bool isSuccessful, DateTimeOffset startTime, TimeSpan duration, Dictionary<string, object> context = null)
+        {
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:335]: We should guard for empty table and account

File: src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs [hunk 7]
```diff
@@ -526,7 +561,7 @@ private static HttpResponse CreateStubResponse(HttpStatusCode statusCode)
 
         private static async Task RetryAssertUntilTelemetryShouldBeAvailableAsync(Func<Task> assertion)
         {
-            await Policy.TimeoutAsync(TimeSpan.FromMinutes(6))
+            await Policy.TimeoutAsync(TimeSpan.FromMinutes(7))
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs:564]: Let's make this configurable

File: docs/features/writing-different-telemetry-types.md (hunk 3 above)
Comment (stijnmoreels) [reply docs/features/writing-different-telemetry-types.md:98]: Thanks we have R#.

File: docs/features/writing-different-telemetry-types.md (hunk 4 above)
Comment (stijnmoreels) [reply docs/features/writing-different-telemetry-types.md:99]: Done.

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs (hunk 6 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:337]: Yes, done. Thanks.

File: src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs]: Made an overload with this timeout.

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs]: Waiting for tests to complete and see if it's improved now

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs]: Should have `Remote dependency name` 🤞

**Comment by stijnmoreels** [comment]
Oef.. @tomkerkhove , build's green.

**Comment by tomkerkhove** [comment]
36m 40s 😜

File: src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs [hunk 8]
```diff
@@ -291,17 +291,58 @@ public static void LogServiceBusDependency(this ILogger logger, string entityNam
...
+        /// <param name="startTime">Point in time when the interaction with the dependency was started</param>
+        /// <param name="duration">Duration of the operation</param>
+        /// <param name="context">Context that provides more insights on the dependency that was measured</param>
+        public static void LogTableStorageDependency(this ILogger logger, string accountName, string tableName, bool isSuccessful, DateTimeOffset startTime, TimeSpan duration, Dictionary<string, object> context = null)
+        {
+            Guard.NotNull(logger, nameof(logger));
+            Guard.NotNullOrWhitespace(tableName, nameof(tableName));
+            Guard.NotNullOrWhitespace(accountName, nameof(accountName));
+
+            logger.LogInformation(DependencyFormat, "Azure Table Storage", tableName, accountName, duration, startTime.ToString(CultureInfo.InvariantCulture), isSuccessful, context);
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs:343]: I was wrong and should actually be this:
```suggestion
            logger.LogInformation(DependencyFormat, "Azure table", tableName, accountName, duration, startTime.ToString(CultureInfo.InvariantCulture), isSuccessful, context);
```

See https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/WEB/Src/DependencyCollector/Shared/Implementation/RemoteDependencyConstants.cs

One option is to add a reference for sake of using these constants but they are internal. Alternative is copying it over but we could then also just stick with our approach.

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/ILoggerExtensions.cs]: Changed and I think we should:
1. Not point to the SDK and stick with our approach
2. But move our dependency types to DependencyTypes.cs which contains all the constants maybe to consolidate it and improve discoverability?

Thoughts @stijnmoreels?

File: src/Arcus.Observability.Tests.Unit/Serilog/ApplicationInsightsTelemetryConverterTests.cs [hunk 9]
```diff
@@ -319,6 +322,54 @@ public void LogServiceBusTopicDependency_WithServiceBusDependency_CreatesDepende
...
+            AssertDoesNotContainLogProperty(logEvent, DependencyTracking.IsSuccessful);
+            AssertDoesNotContainLogProperty(logEvent, DependencyTracking.DependencyData);
+            AssertDoesNotContainLogProperty(logEvent, DependencyTracking.ResultCode);
+            AssertDoesNotContainLogProperty(logEvent, DependencyTracking.StartTime);
+            AssertDoesNotContainLogProperty(logEvent, DependencyTracking.Duration);
+            AssertDoesNotContainLogProperty(logEvent, EventTracking.EventContext);
+            Assert.Collection(telemetries, telemetry =>
+            {
+                var dependencyTelemetry = Assert.IsType<DependencyTelemetry>(telemetry);
+                Assert.Equal("Azure Table Storage", dependencyTelemetry.Type);
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Unit/Serilog/ApplicationInsightsTelemetryConverterTests.cs:362]:
```suggestion
                Assert.Equal("Azure table", dependencyTelemetry.Type);
```

File: src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs [hunk 10]
```diff
@@ -247,17 +238,55 @@ public async Task LogServiceBusDependency_SinksToApplicationInsights_ResultsInSe
...
+                await RetryAssertUntilTelemetryShouldBeAvailableAsync(async () =>
+                {
+                    EventsResults<EventsDependencyResult> results = await client.GetDependencyEventsAsync();
+                    Assert.NotEmpty(results.Value);
+                    Assert.Contains(results.Value, result =>
                     {
-                        EventsResults<EventsDependencyResult> results = await client.GetDependencyEventsAsync();
-                        Assert.NotEmpty(results.Value);
-                        Assert.Contains(results.Value, result => result.Dependency.Type == "Azure Service Bus" && result.Dependency.Target == entityName);
+                        return result.Dependency.Type == "Azure Table Storage"
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Integration/Serilog/ApplicationInsightsSinkTests.cs:284]:
```suggestion
                        return result.Dependency.Type == "Azure table"
```

**Comment by stijnmoreels** [comment]
> 36m 40s 😜

Added #96

**Comment by tomkerkhove** [comment]
We are good to go!

image

**Pull request merged**